harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        # TODO: Check that optimizer is not an amp optimizer.
+        if self.hvd_config.use:
+
+            def filter_named_parameters(
+                model: nn.Module, optimizer: torch.optim.Optimizer,  # type: ignore
+            ) -> List:
+                """
+                filter_named_parameters filters the named parameters of a specified optimizer out
+                of all the named parameters from a specified model. We need this function because
+                a torch.optim.Optimizer doesn't store parameter names and we need the names of
+                the parameters when mapping parameters to each horovod.DistributedOptimizer.
+                """
+                opt_params = {
+                    p for group in optimizer.param_groups for p in group.get("params", [])
+                }
+                return [(name, p) for name, p in model.named_parameters() if p in opt_params]
+
+            use_compression = self.hvd_config.fp16_compression
+            optimizer = hvd.DistributedOptimizer(
+                optimizer,
+                named_parameters=filter_named_parameters(cast(nn.Module, self.model), optimizer),
+                backward_passes_per_step=self.hvd_config.aggregation_frequency,
+                compression=hvd.Compression.fp16 if use_compression else hvd.Compression.none,
+            )
+            logging.debug("Initialized optimizer for distributed and optimized parallel training.")
+
+        self.optimizers.append(optimizer)
+        return optimizer
+
+    def _LRScheduler(
+        self,
+        lr_scheduler: torch.optim.lr_scheduler._LRScheduler,
+        step_mode: pytorch.LRScheduler.StepMode,
+    ) -> pytorch.LRScheduler:
+        # Initialize LR schedulers. This block must happen after wrapping the optimizers using
+        # horovod.DistributedOptimizer because torch.optim._LRScheduler directly access the
+        # optimizer parameter groups, which are also Horovod.DistributedOptimizer relies on.
+        wrapped = pytorch.LRScheduler(lr_scheduler, step_mode)
+        self.lr_schedulers.append(wrapped)
+        return wrapped
+
+    def _init_device(self) -> None:
+        self.n_gpus = len(self.env.container_gpus)
+        if self.hvd_config.use:
+            check.gt(self.n_gpus, 0)
+            # We launch a horovod process per GPU. Each process
+            # needs to bind to a unique GPU.
+            self.device = torch.device(hvd.local_rank())
+            torch.cuda.set_device(self.device)
+        elif self.n_gpus > 0:
+            self.device = torch.device("cuda", 0)
+        else:
+            self.device = torch.device("cpu")
+        check.is_not_none(self.device)
+
+    def _to_device(self, data: pytorch._Data) -> pytorch.TorchData:
+        return pytorch.to_device(
+            data, self.device, self.warning_logged[_WarningLogs.FAILED_MOVING_TO_DEVICE]
+        )
+
+    def _use_amp(self) -> bool:
+        return self.mixed_precision != "O0"
+
+    def _configure_amp(
```

Comment:
question: What do you think about having the name `apex_amp` in this function name?  I'm concerned that if we don't put `apex_amp` in the name, then people will be confused in just a few months when pytorch has their own `amp` as well.

Reply:
I think generally we should be opinionated on what library to use internally rather than giving two choices to users and exposing configuration details to users.

Reply:
Ok, let's consider the possibilities:

If we name this `configure_amp()` (or something non-apex-specific):
- we are allowed to only support apex amp if we think that is good. (good)
- future users might not realize that we are talking about apex amp instead of the torch-native amp. (bad)
- we will have naming problems if we later decide to support apex amp AND torch-native amp.  We can basically guarantee that we will have users who ask for both at some point. (bad)

If we name this `configure_apex_amp()` (or something that is apex-specific):
- we are allowed to only support apex amp if we think that is good. (good)
- future users will know we are talking about apex amp instead of the torch-native amp. (good)
- we will not have any naming problems if we later decide to support apex amp AND torch-native amp. (good)

To me, choosing to name this something non-apex-specific when we know that there's going to be two amp options in the ecosystem very soon (pytorch 1.6) seems like we are opening ourselves up to future problems.

Reply:
okay. @aaron276h What are your thoughts on this name?

Reply:
I would go with `configure_apex_amp()`, that way if we decide to support both methods in the future, the API will make it clean to do so.